iphone/Maps/Core/iCloud/iCloudDocumentsDirectoryMonitor.swift

```diff
@@ -266,4 +228,44 @@ private extension iCloudDocumentsDirectoryMonitor {
       self?.delegate?.didReceiveCloudMonitorError(error)
     }
   }
+
+  // There are no ways to retrieve the content of iCloud's .Trash directory on the macOS because it uses different file system and place trashed content in the /Users/<user_name>/.Trash which cannot be observed without access.
```

Comment:
Is a .Trash or any other folder for deleted files really needed? Why deleted files can't be in the same bookmarks directory, together with non-deleted files?

Reply:
The `.Trash` dir is needed to handle situations when the file was directly deleted from the Files app/ mac.
In this case, we faced up with the absence of the file in the icloud local container and should decide:
1. remove the local file
2. recreate the file in the cloud from the device
In this case, we inspect the cloud's trash to check that file was deleted (because files deleted manually from the Files app are moved to trash).

If there is no file in the cloud's trash the file will be recreated.

Reply:
1. this approach has a huge disadvantage - all the files should be parsed before we can check the flags/attributes. This will lead to the 

> What are the alternative approaches and their pros/cons? E.g.
> 
> 1. Storing "file deleted" flag in the file itself.
> 2. Storing "file deleted" record in a journal file (local or local + iCloud)

1. this approach has a huge disadvantage - all the files should be downloaded and parsed before we can only check the flags/attributes. It will impact performance and mobile data consumption a lot.
2. This approach will allow the tracking of the file's history of changes and not rely on the file's content itself.
pros:
- one source of truth of file changes and we shouldn't rely on the file's `lastModifiedDate` attr directly
- journal is lightweight and fast to sync (like the metadata)
- easier to get the file current state (deleted or not)
cons:
- journal should be synced and merged first to track the file's current state
- if the sync was disabled how to track the changes? This solution will not help to get the
- the journal structure should be well designed and contain enough information to resolve all potential version conflicts but how much info should it contain? 
- should the journal know about the change's content? (Probably not - because it shouldn't replace the kml itself)
-

Reply:
1. Can you please elaborate on details where the overhead will occur? In this approach the latest modified file should always be downloaded and local file should be updated by a remote version, right?

Reply:
> 1. Can you please elaborate on the details of where the overhead will occur? In this approach the latest modified file should always be downloaded and local file should be updated by a remote version, right?

When the icloud's metadata query starts to work and sends the NSMetadataItems list to the app we should decide for every item - download and update the current file or not.

Usually, we download all the files and it's not a blocker.

But in case of **updating** the local file with the cloud's version (or vice versa) we should:
1. download all the cloud's files
2. pass them to the c++ parser (eg files x2: local + cloud) to get the lastModified and deleted/not deleted attributes
3. compare the attributes
4. decide of what how to process files (update/delete/create)
5. process files
6. get a new portion of updates from the cloud with updated metadata and repeat the process from the step 2 until we see that files have equal attributes and up to date.

It may be ok process a lightweight file, but if it will be >5-10mb?
On every bookmark/color update?

But notification comes from the icloud's metadata query asynchronously and **every** event.

We cannot parse files "only once" and forget. As we see the users actively enable/disable sync, uses multiple devices and this addition parsing of the cloud's contents will slow down the app performance.

We should rely on the lightweight information (file's metadata that is already composed by the cloud or journal (custom metadata) to make a decision how to process the file.

Reply:
> In this approach the latest modified file should always be downloaded and local file should be updated by a remote version, right?

It's not straightforward that the remote version is newer.

When the user starts the app or enables sync in the settings we do not have the "previous" state of the local/cloud contents but only the current.

Reply:
Maybe I'm not fully understanding the process. Let's discuss a simple case:

1. Local file was modified (written to). Its modification date was automatically updated by the system.
2. This file was automatically uploaded to iCloud, right?
3. The remote system detected that this file is newer on iCloud (is it feasible?)
4. The remote system downloaded it locally.

Reply:
> Maybe I'm not fully understanding the process. Let's discuss a simple case:

4... on the 2nd device
5.  The local file will be updated from the cloud's version (if there are no version conflicts)
All is correct.

The updation process is quite simple if the current files exist on both devices (except the situation when the sync was disabled and the file was changed on both devices).

Reply:
If this simple case works, then why storing a "deleted" flag inside a file won't work? Isn't it the same write operation as "add a bookmark" or "delete a bookmark"?